core/src/main/scala/api/sfs.scala

```diff
@@ -25,15 +30,19 @@ trait Filesystem {
    *  things too eagerly, nor caching data for too long.
    */
   def resolve(path: Path): Key
-  def metadata(key: Key): Metadata
-  def lookup(key: Key): Data
+  def metadata(key: Key): M[Metadata]
+  def lookup(key: Key): M[Data]
```

Comment:
> All functions require you to return something.

The unix filesystem interface and by extension fuse require you to tell all kinds of lies. There's no way to say "I don't know what the size of this file is" for example, but it's critical not to have to compute every single file in its entirety simply to respond to `ls -l`. I have a bunch of ideas for mitigating this, but it would be a long process because it requires not only out of band data (e.g. extended attributes) but tools which understand it.

However, I'm trying to arrange things here so the metadata returned is explicitly that information which does not require blocking. You should always get an instant response to `ls -l`. The instant response may kick off an asynchronous process which obtains the missing metadata and gives you a better answer the next time, but it isn't going to make you wait for it.

Real life on real filesystems is already like this because (effectively) every file is mutable and might change or disappear at any time. So it's not like this is a step backward. And suffuse-aware programs can take advantage of the aforementioned out-of-band data immediately: they don't have to act as if the size of that file is really -1 bytes or Long.MaxValue or whatever it is which breaks the least stuff.

Playing the cards we're dealt, here.